File: curves.py
from mrcp.turnout import CurveTurnOut_2_3
import mrcp


def fullCurve(panel,pos=(10,10),radius=2):
    curve1=mrcp.Curve(radius=radius,left=True,up=True,color="gray")
    panel.add(curve1,pos)
    curve1=mrcp.Curve(radius=radius,left=False,up=True,color="gray")
    panel.add(curve1,pos=mrcp.movePoint(pos,(2,0)))
    curve1=mrcp.Curve(radius=radius,left=True,up=False,color="gray")
    panel.add(curve1,pos=mrcp.movePoint(pos,(0,2)))
    curve1=mrcp.Curve(radius=radius,left=False,up=False,color="gray")
    panel.add(curve1,pos=mrcp.movePoint(pos,(2,2)))

def r1r4_curves(panel):
    fullCurve(panel,radius=1)
    fullCurve(panel,radius=2)
    fullCurve(panel,radius=3)
    fullCurve(panel,radius=4)

def vertical23(panel):
    curve= CurveTurnOut_2_3(color="gray",up=False,right=True,vertical=True)
    panel.add(curve,(24,12))
    curve= CurveTurnOut_2_3(color="gray",up=False,right=False,vertical=True)
    panel.add(curve,(37,12))
    curve= CurveTurnOut_2_3(color="gray",up=True,right=True,vertical=True)
    panel.add(curve,(24,10))
    curve= CurveTurnOut_2_3(color="gray",up=True,right=False,vertical=True)
    panel.add(curve,(37,10))

def horizontal23(panel):
    curve= CurveTurnOut_2_3(color="gray",up=False,right=True,vertical=False)
    panel.add(curve,(20,24))
    curve= CurveTurnOut_2_3(color="gray",up=True,right=True,vertical=False)
    panel.add(curve,(20,37))
    curve= CurveTurnOut_2_3(color="gray",up=False,right=False,vertical=False)
    panel.add(curve,(18,24))
    curve= CurveTurnOut_2_3(color="gray",up=True,right=False,vertical=False)
    panel.add(curve,(18,37))

def myPanel(name):
    panel = mrcp.Panel(name, 200, 200)
    panel.markStart()

    r1r4_curves(panel)
    vertical23(panel)
    horizontal23(panel)


    panel.paint()
    panel._dwg.save()


if __name__ == '__main__':
    myPanel("curves.svg")

File: mrcp-examples/src/curves.ts
// Copy of curves.py in TS
import {TrackElement,Layout} from "mrcp-layout-model"
import {PainterTco, PainterTcoOptions, TcoPoint,TcoCurveHint} from 'mrcp-painter-tco'


var l:Layout;

const defaultPoint = new TcoPoint({x:10,y:10})

function fullCurve(radius:number=2, pos:TcoPoint=defaultPoint){
    let curve:TrackElement = l.createElement("c1-"+radius);
    curve.addHint(new TcoCurveHint({type:'curve',pos:pos,color:"gray",left:true,up:true,radius:radius}))

    //curve = l.createElement("c2-"+radius);
    //curve.addHint(new TcoCurveHint({type:'curve',pos:pos.move({x:2,y:0}),color:"gray",left:false,up:true,radius:radius}))
    
    //curve = l.createElement("c3-"+radius);
    //curve.addHint(new TcoCurveHint({type:'curve',pos:pos.move({x:0,y:2}),color:"gray",left:true,up:false,radius:radius}))
    
    //curve = l.createElement("c4-"+radius);
    //curve.addHint(new TcoCurveHint({type:'curve',pos:pos.move({x:2,y:2}),color:"gray",left:false,up:false,radius:radius}))

}
    

function r1r4_curves(){
    fullCurve(1)
    //fullCurve(2)
    //fullCurve(3)
    //fullCurve(4)
}

function build_curves_layout(){
    l = new Layout("Test Layout");
    
    r1r4_curves()

}

export async function curves_demo():Promise<void>{
    build_curves_layout();
    console.log("Packed: ")
    console.log(l.toJson())


    let opts:PainterTcoOptions={
        name:"TestCurves",
        format: 'svg',
        outFile:'curves.svg',
        size:{height:200,width:200},
        grid:true
    }

    let panel= new PainterTco(opts)

    panel.grid()
    panel.markStart()

    //r1r4_curves(panel)
    //vertical23(panel)
    //horizontal23(panel)


    panel.paint(l)
    console.log("\nSvg:")
    await panel.save()
    
    return;
}